lib/dal/src/action.rs

```diff
@@ -759,6 +776,142 @@ impl Action {
+            // (this is a backend guard, the button will be hidden from the user)
+            else if Component::resource_by_id(ctx, component_id)
+                .await?
+                .is_none()
+            {
+                Self::enqueue_refresh(ctx, component_id, false).await
+            }
+            // last case, component doesn't exist on head, but has a resource (which can only be true if we ran an import or other mgmt func)
+            // so we can enqueue and dispatch right away
+            else {
+                Self::enqueue_refresh(ctx, component_id, true).await
+            }
+        }
+    }
+
+    async fn enqueue_refresh(
+        ctx: &DalContext,
+        component_id: ComponentId,
+        should_dispatch: bool,
+    ) -> ActionResult<()> {
+        let schema_variant_id = Component::schema_variant_id(ctx, component_id).await?;
+        let refresh_actions = SchemaVariant::find_action_prototypes_by_kind(
+            ctx,
+            schema_variant_id,
+            ActionKind::Refresh,
+        )
+        .await?;
+        if refresh_actions.len() > 1 {
+            return Err(ActionError::UnexpectedNumberOfActionKinds(
+                ActionKind::Refresh,
+                schema_variant_id,
+            ));
+        }
+        if let Some(&prototype_id) = refresh_actions.first() {
+            let maybe_duplicate_action =
+                Action::find_for_kind_and_component_id(ctx, component_id, ActionKind::Refresh)
+                    .await?;
+
+            // if we're not dispatching and we've got a refresh action enqueued already
+            // we'll make this a 'retry' as this whole function exists to interpret what the user wanted
+            // which just means if the existing refresh action is Failed/On Hold, queue it up!
+            if !maybe_duplicate_action.is_empty() && !should_dispatch {
+                if maybe_duplicate_action.len() > 1 {
+                    return Err(ActionError::UnexpectedNumberOfActionsEnqueuedForComponent(
+                        ActionKind::Refresh,
+                        component_id,
+                    ));
+                }
+                if let Some(action_id) = maybe_duplicate_action.first() {
+                    let action = Action::get_by_id(ctx, *action_id).await?;
+                    match action.state() {
+                        ActionState::Failed | ActionState::OnHold => {
+                            Action::set_state(ctx, *action_id, ActionState::Queued).await?
+                        }
+                        ActionState::Dispatched | ActionState::Queued | ActionState::Running => {
+                            return Err(ActionError::ActionAlreadyEnqueued(prototype_id));
+                        }
+                    }
+                }
+            }
+            // if we're dispatching and we've got a refresh action enqueued, check where that action was enqueued from
+            // if it's enqueued in this same change set, dispatch it.
```

Comment:
`If the originating change set is this change set, dispatch it` might be clearer, since it's definitely enqueued in this change set as it's the one were looking at to get the list of queued actions.